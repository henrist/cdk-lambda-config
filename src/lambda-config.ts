import {
  aws_iam as iam,
  aws_lambda as lambda,
  CustomResource,
  custom_resources as cr,
  Duration,
  Stack,
} from "aws-cdk-lib"
import { Construct } from "constructs"
import * as path from "path"
export interface LambdaConfigProps {
  /**
   * The Lambda Function to be updated.
   *
   * If this points to a specific version, the version qualifier
   * will be ignored, but will be used for invalidation instead
   * of providing a separate nonce value.
   *
   * A Lambda Function can only receive a new version based on
   * the latest version.
   */
  function: lambda.IFunction
  /**
   * The configuration to be added to the Lambda Function. Must be
   * a valid JSON structure and cannot contain null values since it
   * is not supported by CloudFormation custom resource properties.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  config: Record<string, any>
  /**
   * Nonce to force new version.
   *
   * If the function provided does not point to a specific
   * version, specify nonce to ensure the function is updated
   * after every change.
   */
  nonce?: string
}

/**
 * Modify a Lambda Function by adding a config.json file using
 * the provided object, and provide a new version for the
 * Lambda Function.
 */
export class LambdaConfig extends Construct {
  public readonly version: lambda.IVersion

  constructor(scope: Construct, id: string, props: LambdaConfigProps) {
    super(scope, id)
    const lambdaConfigProvider = LambdaConfigProvider.getOrCreate(this, {
      locksTableArn: props.config?.locksTable as string,
    })

    const updateCodeResource = new CustomResource(this, "Resource", {
      serviceToken: lambdaConfigProvider.serviceToken,
      properties: {
        FunctionArn: props.function.functionArn,
        Config: props.config,
        Nonce: props.nonce ?? "",
      },
    })

    this.version = lambda.Version.fromVersionArn(
      this,
      "Version",
      updateCodeResource.getAttString("FunctionArn"),
    )
  }
}

interface LambdaConfigProviderProps {
  locksTableArn: string
}

class LambdaConfigProvider extends Construct {
  /**
   * Returns the singleton provider.
   */
  public static getOrCreate(
    scope: Construct,
    props: LambdaConfigProviderProps,
  ) {
    const stack = Stack.of(scope)
    const id = "henrist.lambda-config.provider"
    return (
      (stack.node.tryFindChild(id) as LambdaConfigProvider) ||
      new LambdaConfigProvider(stack, id, props)
    )
  }

  private readonly provider: cr.Provider
  public readonly serviceToken: string

  constructor(scope: Construct, id: string, props: LambdaConfigProviderProps) {
    super(scope, id)
    const statements = [
      new iam.PolicyStatement({
        actions: ["lambda:GetFunction", "lambda:UpdateFunctionCode"],
        resources: ["*"],
      }),
    ]

    if (props.locksTableArn) {
      statements.push(
        new iam.PolicyStatement({
          actions: ["dynamodb:PutItem", "dynamodb:DeleteItem"],
          resources: [props.locksTableArn],
        }),
      )
    }

    this.provider = new cr.Provider(this, "Provider", {
      onEventHandler: new lambda.Function(this, "UpdateCodeFn", {
        code: lambda.Code.fromAsset(path.join(__dirname, "../dist/handler")),
        handler: "index.handler",
        runtime: lambda.Runtime.NODEJS_16_X,
        timeout: Duration.seconds(120),
        initialPolicy: statements,
      }),
    })

    this.serviceToken = this.provider.serviceToken
  }
}
