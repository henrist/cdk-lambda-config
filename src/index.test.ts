import { App, aws_lambda as lambda, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { LambdaConfig } from "./lambda-config"

test("LambdaConfig", () => {
  const app = new App()
  const stack = new Stack(app, "Stack", {
    env: {
      account: "112233445566",
      region: "eu-west-1",
    },
  })

  const fn = new lambda.Function(stack, "Fn", {
    code: lambda.Code.fromInline("code"),
    handler: "index.handler",
    runtime: lambda.Runtime.NODEJS_20_X,
  })

  const fnWithConfig = new LambdaConfig(stack, "LambdaConfig", {
    function: fn,
    config: {
      My: "value",
      Deep: {
        ObjectKey: "Hello world",
      },
    },
  })

  expect(fnWithConfig.version.functionArn).toContain("${Token[TOKEN")
  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})
