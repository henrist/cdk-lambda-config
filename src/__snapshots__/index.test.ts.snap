// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaConfig 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "Fn9270CBC0": Object {
      "DependsOn": Array [
        "FnServiceRoleB9001A96",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "FnServiceRoleB9001A96",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FnServiceRoleB9001A96": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaConfig0334D595": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Config": Object {
          "Deep": Object {
            "ObjectKey": "Hello world",
          },
          "My": "value",
        },
        "FunctionArn": Object {
          "Fn::GetAtt": Array [
            "Fn9270CBC0",
            "Arn",
          ],
        },
        "Nonce": "",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "henristlambdaconfigproviderProviderframeworkonEventB2FC78D4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "henristlambdaconfigproviderProviderframeworkonEventB2FC78D4": Object {
      "DependsOn": Array [
        "henristlambdaconfigproviderProviderframeworkonEventServiceRoleDefaultPolicy3D96FE9E",
        "henristlambdaconfigproviderProviderframeworkonEventServiceRole9AA104AD",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Description": "AWS CDK resource provider framework - onEvent (Stack/henrist.lambda-config.provider/Provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "henristlambdaconfigproviderUpdateCodeFnBF8DF66B",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "henristlambdaconfigproviderProviderframeworkonEventServiceRole9AA104AD",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "henristlambdaconfigproviderProviderframeworkonEventServiceRole9AA104AD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "henristlambdaconfigproviderProviderframeworkonEventServiceRoleDefaultPolicy3D96FE9E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "henristlambdaconfigproviderUpdateCodeFnBF8DF66B",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "henristlambdaconfigproviderProviderframeworkonEventServiceRoleDefaultPolicy3D96FE9E",
        "Roles": Array [
          Object {
            "Ref": "henristlambdaconfigproviderProviderframeworkonEventServiceRole9AA104AD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "henristlambdaconfigproviderUpdateCodeFnBF8DF66B": Object {
      "DependsOn": Array [
        "henristlambdaconfigproviderUpdateCodeFnServiceRoleDefaultPolicyE51A664A",
        "henristlambdaconfigproviderUpdateCodeFnServiceRole2FF24712",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "henristlambdaconfigproviderUpdateCodeFnServiceRole2FF24712",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "henristlambdaconfigproviderUpdateCodeFnServiceRole2FF24712": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "henristlambdaconfigproviderUpdateCodeFnServiceRoleDefaultPolicyE51A664A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "lambda:GetFunction",
                "lambda:UpdateFunctionCode",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "henristlambdaconfigproviderUpdateCodeFnServiceRoleDefaultPolicyE51A664A",
        "Roles": Array [
          Object {
            "Ref": "henristlambdaconfigproviderUpdateCodeFnServiceRole2FF24712",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
